### PR TITLE
Phase5 controller

### PR DESCRIPTION
**Pull request by cxoltero**

### CODE REVIEW

File: src/js/controllers/imageCtrl-test.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        $q = _$q_;
+
+        vm = $controller('imagesCtrl');
+
+    }));
+
+    describe('imagesCtrl', function(){
+        beforeEach(function () {
+            images.loadPics.returns($q.resolve(
+                {
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: Can we move this return object to a globally available variable? That way any test can reference, if needed. You'll need to declare it with your other top-level variables, before the first `beforeEach()`.

File: src/js/controllers/imageCtrl-test.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                            title: ""
+                        }
+                        ]
+                    }
+                }
+            ));
+        });
+
+        beforeEach(
+            function(){
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: Let's keep the callback on the same line as the beforeEach.

File: src/js/controllers/imageCtrl-test.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                vm.getImages();
+                $rootScope.$apply();
+            }
+        );
+
+        describe('returned flickr images object from http call', function(){
+
+            it('should exist', function () {
+                expect(vm.getImages).to.exist();
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: This is a good test, however, I worry that we're actually calling the method (in the `beforeEach()`), before we actually check if it exists. I'd suggest just moving this test to its own `describe()` above so it is not impacted by the `vm.getImages()` call.

File: src/js/controllers/imageCtrl-test.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    }));
+
+    describe('imagesCtrl', function(){
+        beforeEach(function () {
+            images.loadPics.returns($q.resolve(
+                {
+                    contentType: "text/json",
+                    status: "success",
+                    photos: {
+                        photo: [
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: I'm not familiar with the Flickr API responses, but this looks strange. **Photos** is an object, and **photo** is an array. Seems to me that the array should have the plural label. Can you confirm this response object is correct?

File: src/js/controllers/imageCtrl.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+(function (angular) {
+    angular
+        .module('flickrPOC')
+        .controller('imagesCtrl', imagesCtrl);
+
+    imagesCtrl.$inject = ['images'];
+
+    function imagesCtrl(images) {
+        /* jshint validthis: true */
+        var vm = this;
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: I understand why the `validthis` rule needs to be disabled. JSHint doesn't like the reference to `this`, and although the general override is effective, it disables the rule for _all lines_ that follow it. This can cause issues where invalid code isn't discovered by the linter. What is preferred is simply disabling the rule on a line-by-line basis, as required. 

So, this:

``` javascript
/* jshint validthis: true */
var vm = this;
```

Becomes this:

``` javascript
var vm = this; // jshint ignore:line
```

The single line ignore allows other developers to know exactly what line is causing the linter to complain, thus can feel confident that the rest of the file is tip-top.

File: src/js/controllers/imageCtrl.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+        vm.getImages = function () {
+            images.loadPics()
+                .then(function (res) {
+                    //console.log('this is not from test '+ res);
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: Remove comment.

File: src/js/controllers/imageCtrl.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+                    function(err){
+                        console.error('Error loading images ', err.status, err.data);
+                    })
+                .finally(function () {
+                    console.log("finally finished");
+                });
+
+        };
+        function activate(){
+            vm.getImages();
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: Are you planning on using `getImages()` in the HTML on a button click or similar? If so, then no worries. But, if not, and `vm.getImages()` only exists to be called internally by the controller, then we can remove the method from the `vm` and simply make the `images.loadPics()` call in the `activate()`function.

``` javascript
function activate(){
  images.loadPics()
    .then()
    .catch()
}
```

This change will require some updates to the way the tests are run. See my comments on that file for details.

File: src/js/controllers/imageCtrl.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                    //console.log('this is not from test '+ res);
+                    vm.images = res;
+                    return vm.images;
+                })
+                .catch(
+                    function(err){
+                        console.error('Error loading images ', err.status, err.data);
+                    })
+                .finally(function () {
+                    console.log("finally finished");
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: No need for finally() if only logging a message.

File: src/js/controllers/imageCtrl.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        vm.getImages = function () {
+            images.loadPics()
+                .then(function (res) {
+                    //console.log('this is not from test '+ res);
+                    vm.images = res;
+                    return vm.images;
+                })
+                .catch(
+                    function(err){
+                        console.error('Error loading images ', err.status, err.data);
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: Please use $log to log messages to the console. It allows for tests to be easily ran against the messages.

File: src/js/controllers/imageCtrl-test.js (hunk 4 above)
Comment (cxoltero) [inline src/js/controllers/imageCtrl-test.js]: Yeah, photo shows up as an array of objects. 
screen shot 2016-01-07 at 11 01 31 am

File: src/js/controllers/imageCtrl.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+                .finally(function () {
+                    console.log("finally finished");
+                });
+
+        };
+        function activate(){
+            vm.getImages();
+        }
+
+        //activate();
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: Looks like activate is commented out. This would prevent the controller from triggering the data call. Please uncomment so the tests are ran against the live controller.

File: src/js/directives/image-list-directive.test.js [hunk 11]
```diff
@@ -2,133 +2,131 @@
...
         });
 
-      });
-
-  });
-
-});
+    });
 
+});
```
Comment (paul-barry-kenzan) [inline src/js/directives/image-list-directive.test.js]: These changes shouldn't be part of this PR. Please remove and create an additional PR if these changes are required.

File: src/js/services/getImgService.js [hunk 12]
```diff
@@ -4,44 +4,43 @@
...
-                return images;
+                images.push(picUrl);
             }
 
+            return images;
         }
-})(angular);
 
+    }
+})(angular);
```
Comment (paul-barry-kenzan) [inline src/js/services/getImgService.js:46]: Same goes for these changes. If they aren't required for the controller updates, please create a new PR for these updates.

File: src/js/services/getImgService.test.js [hunk 13]
```diff
@@ -121,4 +121,4 @@ describe('getImagesService', function(){
             });
         });
     });
-});
+});
```
Comment (paul-barry-kenzan) [inline src/js/services/getImgService.test.js]: Same here. This should be part of a separate PR.

File: src/js/directives/image-list-directive.js [hunk 14]
```diff
@@ -51,21 +51,4 @@
...
-
- */
-
-/*
- View Controller
- Retrieve the images via Service
- Store images array in VM property
- VM property passed to directive. <-- ARRAY
-
- */
```
Comment (paul-barry-kenzan) [inline src/js/directives/image-list-directive.js:71]: Same here. This should be part of a separate PR.

File: src/js/controllers/imageCtrl.js (hunk 7 above)
Comment (cxoltero) [inline src/js/controllers/imageCtrl.js]: no, I just wanted the images to load once the page loads. in this case, should we just remove the activate method? or set it to when the page loads?

Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: @cxoltero 
So what I'd like you to do is move the `images.loadImages()` to the `activate()` function, allowing the `vm.getImages()` function to be deleted. This way, the controller will load, then call `activate()`, which triggers the image retrieval. 

In your tests, instead of calling `vm.getImages()` to get the images, you'll just have to instantiate the controller for each test, which you already do on line#50 (`vm = $controller('imagesCtrl');') of the test file. The reason you'll have to do that for each test is to ensure that your values are created, and not having to depend on previous tests.

**Comment by paul-barry-kenzan** [comment]
@cxoltero 
A general comment about file naming. We haven't discussed it specifically, but we'll need to keep a standard across all the files. Right now, the directive is using **snake-case**, while other files are using **camelCase**. Once this PR gets merged, could you submit one that standardizes the file names? Use John Papa's guide as a reference and come to me for any questions.

File: src/js/controllers/imageCtrl-test.js [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+        images = _images_;
+        $q = _$q_;
+
+        vm = $controller('imagesCtrl');
+
+    }));
+
+    describe('imagesCtrl', function(){
+        beforeEach(function () {
+            images.loadPics.returns($q.resolve(mockImageObject));
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: MUCH better! No arguing what this call returns.

File: src/js/controllers/imageCtrl.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+        function activate(){
+
+            images.loadPics()
+                .then(function (res) {
+                    vm.images = res;
+                    return vm.images;
+                })
+                .catch(
+                    function(err){
+                        $log.error('Error loading images ', err.status, err.data);
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: 2 things:
1. Can you keep the callback on the same line as `.catch(`?
2. Can you add test(s) for the $log call?

**Comment by paul-barry-kenzan** [comment]
@cxoltero 
I think this PR will benefit from a rebase of Master. That should remove the directive and service changes.

File: src/js/controllers/imageCtrl-test.js [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+            }));
+            beforeEach(inject(function (_$rootScope_, _$controller_, _images_, _$q_, _$log_) {
+                $rootScope = _$rootScope_.$new();
+                $controller = _$controller_;
+                images = _images_;
+                $q = _$q_;
+                $log = _$log_;
+                vm = $controller('imagesCtrl');
+
+                images.loadPics.returns($q.resolve(mockImageObject));
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: This isn't needed since you're initializing the return value in the $provide block.

File: src/js/controllers/imageCtrl-test.js [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+                it('check for neccesary attibutes to not be null ', function(){
+                    angular.forEach(photo, function(value){
+                        expect(value.farm).not.to.be.null();
+                        expect(value.server).not.to.be.null();
+                        expect(value.id).not.to.be.null();
+                        expect(value.owner).not.to.be.null();
+                        expect(value.secret).not.to.be.null();
+                    });
+                });
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js:99]: The 3 tests are ultimately testing different variations of the same thing. I agree that a check to ensure that the values exist is useful, however, I feel 2 of these are overkill. :smile: 

If you'd like test something in addition to the existence, you could check the specific values of one (or more) of the returned data. And if you plan on doing any data manipulation, you could verify that the values match the appropriate format.

File: src/js/controllers/imageCtrl-test.js [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+                $q = _$q_;
+                $log = _$log_;
+                vm = $controller('imagesCtrl');
+
+                $rootScope.$apply();
+            }));
+
+            it('Should log error message ', function(){
+                var errorMsg = 'Error loading images ';
+                expect($log.error.logs[0]).to.include(errorMsg);
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: Not a bad way to test whether the $log has been called. Typically, I use a spy here and check what against the string value of the message.

File: src/js/controllers/imageCtrl.js [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+
+    function imagesCtrl(images, $log) {
+        var vm = this;// jshint ignore:line
+
+        function activate(){
+
+            images.loadPics()
+                .then(function (res) {
+                    vm.images = res;
+                    return vm.images;
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js]: Is this statement needed to pass your tests? I don't think it should be. In this case, we're not really looking to return a value. Once we set the value to `vm.images`, Angular's data-binding will update any uses. If this is required to pass tests, let's take a look together.

File: src/js/controllers/imageCtrl.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+                .catch(function(err){
+                    $log.error('Error loading images ', err.status, err.data);
+                });
+        }
+
+        activate();
+
+    }
+
+})(angular);
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js:28]: Good looking controller. Very little code, no complex logic, and testable. Great job.

Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl.js:28]: One additional note, look at the difference in number of lines of code between the Controller (27) and the Tests (142). It may seem like that tests are overkill and would take up too much time to write and maintain, however, what I see is a solid controller that I can feel comfortable using and modifying.

File: src/js/controllers/imageCtrl-test.js [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+
+            });
+
+        });
+
+        describe('Error from to api', function() {
+
+            beforeEach(module(function($provide){
+                $provide.service('images', function($q, $log){
+                    this.loadPics = sinon.stub().returns($q.reject($log));
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: This doesn't look right. The loadPics method wouldn't be returning the $log service, but an error object of sorts. The $log service should be used to log the message, but the return value from the loadPics method itself should be an object.

Comment (cxoltero) [inline src/js/controllers/imageCtrl-test.js]: Here we would pass an empty object and get the response to add content to it. Right?

File: src/js/controllers/imageCtrl-test.js [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+
+            });
+
+        });
+
+        describe('Error from to api', function() {
+
+            beforeEach(module(function($provide){
+                $provide.service('images', function($q){
+                    this.loadPics = sinon.stub().returns($q.reject({}));
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: What should be returned is an object the matches what the $http.get() would return on a failure. One way to test would be to attempt to ping the Flickr server with a wrong URL or UserID and see what is returned. If that doesn't seem to work, an generic error message should suffice.

``` javascript
var mockError = {
  status: 500,
  statusText: 'Server error'
}
```

All we're really trying to test, is that when there's an error in the response of Flickr, the controller handles that properly. The specific message isn't important per se.

File: src/js/controllers/imageCtrl-test.js [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+                   });
+               });
+
+            });
+
+        });
+
+        describe('Error from to api', function() {
+            var errorObject = { "stat": "fail", "code": 2, "message": "Unknown user" }
+            var errorMsg = 'Unknown user';
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: The `errorObject` is much better. Nice work. And I see what you're doing with the additional `errorMsg`. I have one suggestion though.

When you're working with mock data, you can _almost_ always assume that someone will have to add to the data at some point in the future to handle specific use cases. To safe guide against this, I tend to use the mock object properties when asserting a value against the mock data.

For example, you have a test below: 

``` javascript
it('Should log error message ', function(){
  expect($log.error.logs[0][0]).to.include(errorMsg);
});
```

This test is a valuable one to ensure that the $log service is called with the provided error message. No argument there. But what if someone changes the `errorObject.message` string, but overlooks the second reference to the message string (`errorMsg`)? An additional change would need to be made to match, duplicating the string. Not the best practice.

Instead, consider using the `errorObject` itself in your assertion, ensuring that any changes made will retain the passing test.

``` javascript
it('Should log error message ', function(){
  expect($log.error.logs[0][0]).to.include(errorObject.message);
});
```

File: src/js/controllers/imageCtrl-test.js [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+                images = _images_;
+                $q = _$q_;
+                $log = _$log_;
+                vm = $controller('imagesCtrl');
+
+                $rootScope.$apply();
+            }));
+
+            it('Should log error message ', function(){
+                expect($log.error.logs[0][0]).to.include(errorMsg);
```
Comment (paul-barry-kenzan) [inline src/js/controllers/imageCtrl-test.js]: Consider using the following to avoid the array of arrays reference (`[0][0]`):

``` javascript
var spy = sinon.spy($log, 'error');
expect(spy).to.have.been.calledWith(errorObject.message);
```

**Comment by paul-barry-kenzan** [comment]
@cxoltero 
This looks good. You can merge, and rebase your implementation branch
